crates/autopilot/src/decoded_settlement.rs

```diff
@@ -197,16 +175,68 @@ impl TryFrom<database::orders::OrderExecution> for OrderExecution {
 }
 
 impl DecodedSettlement {
+    /// Number of bytes that may be appended to the calldata to store an auction
+    /// id.
+    pub const META_DATA_LEN: usize = 8;
+
     pub fn new(input: &[u8]) -> Result<Self, DecodingError> {
         let function = GPv2Settlement::raw_contract()
             .abi
             .function("settle")
             .unwrap();
-        let decoded_input = decode_function_input(function, input)?;
-        <DecodedSettlementTokenized>::from_token(Token::Tuple(decoded_input))
-            .map(Into::into)
-            .context("failed to decode settlement")
-            .map_err(From::from)
+        let without_selector = input
+            .strip_prefix(&function.selector())
+            .ok_or(DecodingError::InvalidSelector)?;
+
+        // Decoding calldata without expecting metadata can succeed even if metadata
+        // was appended. The other way around would not work so we do that first.
+        if let Ok(decoded) = Self::try_new(without_selector, function, true) {
+            return Ok(decoded);
+        }
+        Self::try_new(without_selector, function, false).map_err(Into::into)
```

Comment:
I'm slightly confused that we have to try once with and once without metadata. The way I would expect it it to work is that parsing the calldata with the function's ABI tells you how many bytes were consumed. The leftover bytes must then be the metadata. Maybe we don't have the right APIs for this and maybe this is related to your comment below.

Reply:
> The way I would expect it it to work is that parsing the calldata with the function's ABI tells you how many bytes were consumed.

Yeah, this would have been my preferred solution as well (this [comment](https://github.com/cowprotocol/ethcontract-rs/pull/936#issuecomment-1630558317) had exactly that in mind). But unless I'm missing something this functionality can not be patched using `ethcontract-rs` and I didn't want to copy lots of code over from `ethabi` to know exactly how many bytes would be used by that implementation.

Reply:
Note that Solidity ABI allows for arbitrary padding - so even then I don't think this would be "correct".

Reply:
Totally agree, with that assessment. (That's why I suggested the idea of generating a secret key in the first place during the design phase).
Not sure why it we switched over to using the `auction_id`.

Reply:
It seamed reasonable at the time. I think it was because this was the simplest solution. See https://cowservices.slack.com/archives/C049P351GKF/p1688227366128809.


I think we can release with `settle(...) || auctionId` and can switch to `settle(...) || salt`.

Reply:
> I think it was because this was the simplest solution.

Strange argument given the fact that half of the logic was already there.